ipfssync/subpub/stream.go

```diff
@@ -0,0 +1,60 @@
+package subpub
+
+import (
+	"bufio"
+	"fmt"
+
+	"github.com/libp2p/go-libp2p-core/network"
+	libpeer "github.com/libp2p/go-libp2p-core/peer"
+	"go.vocdoni.io/dvote/log"
+)
+
+func (ps *SubPub) handleStream(stream network.Stream) {
+	// 'stream' will stay open until you close it (or the other side closes it).
+
+	peer := stream.Conn().RemotePeer()
+	log.Infof("connected to peer %s: %+v", peer, stream.Conn().RemoteMultiaddr())
+
+	// Any messages read from the stream are sent to the SubPub.Reader channel.
+	go ps.readHandler(stream)
+
+	// Create a buffer stream for non blocking write.
+	ps.Unicasts[peer] = bufio.NewWriter(stream)
+}
+
+func (ps *SubPub) Unicast(address string, message []byte) error {
+	peerID, err := libpeer.Decode(address)
+	if err != nil {
+		return fmt.Errorf("cannot decode %s into a peerID: %w", address, err)
+	}
+	log.Debugf("sending %d bytes to %s = %s", len(message), address, peerID)
+	if err := ps.SendMessage(ps.Unicasts[peerID], message); err != nil {
+		return fmt.Errorf("error trying to unicast to %s: %v", peerID, err)
+	}
+	return nil
+}
+
+func (ps *SubPub) readHandler(stream network.Stream) {
+	r := bufio.NewReader(stream)
+	peer := stream.Conn().RemotePeer().String()
+
+	// Ensure that we always close the stream.
+	defer stream.Close()
+
+	for {
+		select {
+		case <-ps.close:
+			return
+		default:
+			// continues below
+		}
+		message, err := ps.ReadMessage(r)
+		if err != nil {
+			log.Errorf("error reading stream from %s: %v", peer, err)
+			return
+		}
+		message.Peer = peer
+
+		go func() { ps.Reader <- message }()
```

Comment:
not sure if we got this from the old code, but this goroutine feels wrong for two reasons:

1) since we spawn one goroutine per incoming message to pass it along, we may end up seeing the messages on the other side out of order. even if we don't technically care about the order, this could lead to a message not being seen for minutes or hours just because we have lots of messages coming in that jump the queue.

2) it's a potential DoS attack vector; I can send a libp2p peer a million tiny messages all at once, and since each will spawn a goroutine which takes up ~2KiB on the stack, then that will easily mean a 2GiB spike in memory usage.

both problems go away if we remove the goroutine. messages will be kept in order, so the oldest gets priority, and if messages come in so fast that we can't handle them, then the network layer would handle that for us - the writer on the other side would block.

Reply:
indeed old code, but absolutely insightful comments <3